### PR TITLE
test(logger): fix logger tests failing when running in WebStorm

### LINKED ISSUE #225: Logger tests not working when executing them in WebStorm
jest default reporter override `process.stdout.write` https://github.com/facebook/jest/blob/37200eb7cb0abac55801d32ed439f2925df3899e/packages/jest-reporters/src/DefaultReporter.ts#L100-L104

Pino checks if the `process.stdout.write` has been tempered, in case it is it uses it, if not it uses the `sonic-boom` which writes to `process.stdout.fd`
https://github.com/pinojs/pino/blob/dc4e8546c0a4bedb087cbcccd9a48b56100b0ad7/lib/tools.js#L305-L311

when starting from WebStorm, WebStorm uses its own jest reporter which doesn't override the stdout which causes the issue

### PR DESCRIPTION
**Pull request by rluvaton**
Fix #225

### CODE REVIEW

**Comment by rluvaton** [comment]
Why the project coverage dropped? I haven't changed the actual tests

File: src/code-templates/libraries/logger/test/index.test.ts [hunk 1]
```diff
@@ -7,77 +7,71 @@ beforeEach(() => {
...
-  test('When no explicit configuration is set, info logs are written', async () => {
-    // Arrange
-    const stdoutStub = sinon.stub(process.stdout, 'write');
+  let stdoutStub: sinon.SinonStubbedMember<typeof process['stdout']['write']>;
 
+  beforeEach(() => {
+    // Must stub the process.stdout before we configure the logger
+    // as it would fail when the underlying logger is pino, and we run in WebStorm
+    // See more here - practicajs/practica#225
+    stdoutStub = sinon.stub(process.stdout, 'write');
```
Comment (goldbergyoni) [inline src/code-templates/libraries/logger/test/index.test.ts:16]: Why not keep this within the tests? As a general rule of thumb, anything that directly affects the test is much more readable when its part of the test story (vs external side effect)

File: src/code-templates/libraries/logger/test/index.test.ts [hunk 2]
```diff
@@ -7,77 +7,71 @@ beforeEach(() => {
...
+
+  test('When no explicit configuration is set, info logs are written', async () => {
     // Act
     logger.info('This is an info message');
 
     // Assert
-    expect({ stdCallCount: stdoutStub.callCount }).toMatchObject({
-      stdCallCount: 1,
-    });
-    const lastStdoutCall = JSON.parse(stdoutStub.lastCall?.firstArg);
```
Comment (goldbergyoni) [inline src/code-templates/libraries/logger/test/index.test.ts:21]: I did this intentionally so the test result will look like:

expected {stdCallCount:1) but got {stdCallCount:0) and the tests result will not look like:

expected 1 but got 0

File: src/code-templates/libraries/logger/test/index.test.ts [hunk 3]
```diff
@@ -7,77 +7,71 @@ beforeEach(() => {
...
     // Arrange
     logger.configureLogger({ level: 'error' }, true);
-    const stdoutStub = sinon.stub(process.stdout, 'write');
 
     // Act
     logger.info('This is an info message');
 
     // Assert
-    expect(stdoutStub.callCount).toBe(0);
+    expect(stdoutStub.callCount).toEqual(0);
```
Comment (goldbergyoni) [inline src/code-templates/libraries/logger/test/index.test.ts:64]: Good call 💪

File: src/code-templates/libraries/logger/test/index.test.ts [hunk 4]
```diff
@@ -7,77 +7,71 @@ beforeEach(() => {
 });
 
 describe('logger', () => {
-  test('When no explicit configuration is set, info logs are written', async () => {
-    // Arrange
-    const stdoutStub = sinon.stub(process.stdout, 'write');
+  let stdoutStub: sinon.SinonStubbedMember<typeof process['stdout']['write']>;
 
+  beforeEach(() => {
```
Comment (goldbergyoni) [inline src/code-templates/libraries/logger/test/index.test.ts:12]: Do we really need two hirearchical beforeEach? One is simpler

**Review by goldbergyoni — CHANGES_REQUESTED** [review]
@rluvaton Webstorm developers will thank you for this

Q - Isn't stubbing at the begining of the test good enough (before logger is called)? why do we need to put it inside an externous hook?

**Comment by rluvaton** [comment]
It is but I didn't want to duplicate the comment...

File: src/code-templates/libraries/logger/test/index.test.ts (hunk 2 above)
Comment (rluvaton) [reply src/code-templates/libraries/logger/test/index.test.ts:21]: This is why I prefer using jest spies as there matchers that much more helpful than doing this as it gives you more context of the actual argument.

Maybe we should use sinon matchers for jest?

File: src/code-templates/libraries/logger/test/index.test.ts (hunk 3 above)
Comment (rluvaton) [reply src/code-templates/libraries/logger/test/index.test.ts:64]: thanks

File: src/code-templates/libraries/logger/test/index.test.ts (hunk 4 above)
Comment (rluvaton) [reply src/code-templates/libraries/logger/test/index.test.ts:12]: agree

File: src/code-templates/libraries/logger/test/index.test.ts (hunk 2 above)
Comment (goldbergyoni) [reply src/code-templates/libraries/logger/test/index.test.ts:21]: Yes!

p.s. Jest mocking is a world full with magics, this is why I avoid it all together (even the good parts)

File: src/code-templates/libraries/logger/test/index.test.ts (hunk 1 above)
Comment (rluvaton) [reply src/code-templates/libraries/logger/test/index.test.ts:14]: Agree but duplicating this multiple times will make the test less readable in my opinion

File: src/code-templates/libraries/logger/test/index.test.ts (hunk 2 above)
Comment (rluvaton) [reply src/code-templates/libraries/logger/test/index.test.ts:21]: reverted back to use stdCallCount

File: src/code-templates/libraries/logger/test/index.test.ts [hunk 5]
```diff
@@ -1,31 +1,34 @@
...
-beforeEach(() => {
-  sinon.restore();
-  logger.resetLogger();
-});
-
 describe('logger', () => {
-  test('When no explicit configuration is set, info logs are written', async () => {
-    // Arrange
-    const stdoutStub = sinon.stub(process.stdout, 'write');
+  let stdoutStub: sinon.SinonStubbedMember<typeof process['stdout']['write']>;
```
Comment (goldbergyoni) [inline src/code-templates/libraries/logger/test/index.test.ts:5]: @rluvaton 

I'm concerned with this line as it raises the reader's reading complexity, is there any value here?

**Review by goldbergyoni — APPROVED** [review]
@rluvaton Approved and left few comments for you to consider

File: src/code-templates/libraries/logger/test/index.test.ts (hunk 1 above)
Comment (rluvaton) [reply src/code-templates/libraries/logger/test/index.test.ts:14]: less readable as every time you will see a long comment which will make the test longer and people will just ignore it, instead, I spied in every function and added the comment only to the first stub

File: src/code-templates/libraries/logger/test/index.test.ts (hunk 5 above)
Comment (rluvaton) [reply src/code-templates/libraries/logger/test/index.test.ts:5]: type safety, removed and recreate in every test

**Review by goldbergyoni — APPROVED** [review]

**Comment by goldbergyoni** [comment]
@rluvaton Just pinging to ensure you've noticed that I approved

**Pull request merged**